Make postMessage work for site-isolation

chirags27 opened the pull request:
#### 5eb3a34b116bdda06cc92e3b2c5c289ceb82074a
<pre>
Make postMessage work for site-isolation
<a href="https://bugs.webkit.org/show_bug.cgi?id=253588">https://bugs.webkit.org/show_bug.cgi?id=253588</a>
rdar://106439413

Reviewed by Alex Christensen.

This change makes postMessage ping pong work by creating the Frame with
the correct identifiers and remote processIDs so that message sending
and reception between cross origin iframes can work with postMessage.

* Source/WebCore/page/DOMWindow.cpp:
(WebCore::DOMWindow::createTargetOriginForPostMessage):
* Source/WebCore/page/DOMWindow.h:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::processPostMessage):
(WebCore::LocalDOMWindow::postMessage):
(WebCore::LocalDOMWindow::postMessageFromRemoteFrame):
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/page/Page.cpp:
(WebCore::createMainFrame):
* Source/WebCore/page/PageConfiguration.cpp:
(WebCore::PageConfiguration::PageConfiguration):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebCore/page/RemoteDOMWindow.cpp:
(WebCore::RemoteDOMWindow::postMessage):
* Source/WebCore/page/RemoteDOMWindow.h:
* Source/WebCore/page/RemoteFrame.cpp:
(WebCore::RemoteFrame::createMainFrame):
(WebCore::RemoteFrame::createSubframe):
(WebCore::RemoteFrame::createSubframeWithContentsInAnotherProcess):
(WebCore::RemoteFrame::RemoteFrame):
(WebCore::m_remoteProcessIdentifier):
(WebCore::m_layerHostingContextIdentifier): Deleted.
* Source/WebCore/page/RemoteFrame.h:
* Source/WebCore/page/RemoteFrameClient.h:
* Source/WebKit/Shared/FrameTreeCreationParameters.h:
* Source/WebKit/Shared/FrameTreeCreationParameters.serialization.in:
* Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
(WebKit::ProvisionalFrameProxy::ProvisionalFrameProxy):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::commitProvisionalFrame):
(WebKit::WebFrameProxy::frameTreeCreationParameters const):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::postMessageToRemote):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::postMessageToRemote):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::createRemoteSubframe):
(WebKit::WebFrame::didCommitLoadInAnotherProcess):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::clientForMainFrame):
(WebKit::m_appHighlightsVisible):
(WebKit::WebPage::constructFrameTree):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::remotePostMessage):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/262334@main">https://commits.webkit.org/262334@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cacacbd51ab51b296d9b23d4fd3921918ab0bb2b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1193 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1225 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1264 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1936 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1065 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1176 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1276 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1296 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1237 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1201 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1114 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1098 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1807 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1141 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1096 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1082 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1089 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1126 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2181 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1137 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1046 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1075 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1105 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1162 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/127 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -324,7 +324,7 @@ void WebFrame::didCommitLoadInAnotherProcess(WebCore::LayerHostingContextIdentif
     coreFrame->disconnectOwnerElement();
     auto client = makeUniqueRef<WebRemoteFrameClient>(*this, WTFMove(invalidator));
-
+    m_frameID = WebCore::FrameIdentifier(m_frameID.object(), processIdentifier);
```
This is incorrect.  We want the FrameIdentifier to have the process identifier of the process that generated the identifier.

chirags27 replied:
Talking offline about how this makes sense, because that almost seems like redundant information to keep with the RemoteFrame. If it's required, I could investigate other ways of getting this information to RemoteFrame.

achristensen07 commented on Source/WebCore/page/RemoteFrame.cpp:
```diff
@@ -51,7 +51,7 @@ Ref<RemoteFrame> RemoteFrame::createSubframeWithContentsInAnotherProcess(Page& p
 RemoteFrame::RemoteFrame(Page& page, UniqueRef<RemoteFrameClient>&& client, FrameIdentifier frameID, HTMLFrameOwnerElement* ownerElement, AbstractFrame* parent, Markable<LayerHostingContextIdentifier> layerHostingContextIdentifier)
     : AbstractFrame(page, frameID, ownerElement, parent)
-    , m_window(RemoteDOMWindow::create(*this, GlobalWindowIdentifier { Process::identifier(), WindowIdentifier::generate() }))
+    , m_window(RemoteDOMWindow::create(*this, GlobalWindowIdentifier { frameID.processIdentifier(), WindowIdentifier::generate() }))
```
I'm not sure about this change.

chirags27 replied:
Updated based on the offline conversation

chirags27 commented on Source/WebCore/page/DOMWindow.cpp:
```diff
@@ -44,8 +44,8 @@ HashMap<GlobalWindowIdentifier, DOMWindow*>& DOMWindow::allWindows()
 {
-    ASSERT(!allWindows().contains(identifier));
-    allWindows().add(identifier, this);
+    ASSERT(!allWindows().contains(m_identifier));
```
This looks like a pretty serious bug that should be fixed

chirags27 replied:
Fixed in a separate PR

chirags27 commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -983,6 +983,61 @@ ExceptionOr<void> LocalDOMWindow::postMessage(JSC::JSGlobalObject& lexicalGlobal
     return { };
 }
 
+ExceptionOr<void> LocalDOMWindow::postMessageFromRemote(JSC::JSGlobalObject& lexicalGlobalObject, LocalDOMWindow& incumbentWindow, std::optional<WebCore::SecurityOriginData> target, const WebCore::MessageWithMessagePorts& message)
```
This is copied from @pascoej's PR to make sure postMessage works.

pascoej commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1315,6 +1315,13 @@ void WebProcessProxy::didDestroyUserGestureToken(uint64_t identifier)
         m_userInitiatedActionByAuthorizationTokenMap.remove(*removed->authorizationToken());
 }
 
+void WebProcessProxy::postMessageToRemote(WebCore::ProcessIdentifier destinationProcessIdentifier, WebCore::FrameIdentifier identifier, std::optional<WebCore::SecurityOriginData> target, const WebCore::MessageWithMessagePorts& message)
```
This applies all over the patch, but we can get the ProcessIdentifier of the frame from `frameIdentifier.processIdentifier()`

chirags27 replied:
Yeah, the issue is that the frameIdentifier doesn't have the right processIdentifier. FrameIdentifier always has the processIdentifier for the process which is created the frame object (maybe it be LocalFrame or RemoteFrame), but not which is hosting the frame.

achristensen07 commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -983,6 +983,60 @@ ExceptionOr<void> LocalDOMWindow::postMessage(JSC::JSGlobalObject& lexicalGlobal
+            }
+        }
+
+        auto* globalObject = document()->globalObject();
```
Let's null check the document too, just in case.  If there's no document, do nothing.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -56,14 +56,13 @@ static void enableSiteIsolation(WKWebViewConfiguration *configuration)
     auto webkitHTML = "<script>"
     "    window.addEventListener('message', (event) => {"
-    "        parent.window.postMessage('pong', '*')"
+    "           parent.window.postMessage(event.data + 'pong', '*');"
```
Don't add 3 spaces of indentation before "parent"

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -139,7 +138,112 @@ HTTPServer server(HTTPServer::UseCoroutines::Yes, [&](Connection connection) ->
+    "        alert('parent frame received ' + event.data)"
+    "    };"
+    "    onload = () => {"
+    "        document.getElementById('webkit_frame').contentWindow.postMessage('ping', '*', [channel.port2,]);"
```
extra comma

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -139,7 +138,112 @@ HTTPServer server(HTTPServer::UseCoroutines::Yes, [&](Connection connection) ->
+                co_await connection.awaitableSend(makeString("HTTP/1.1 200 OK\r\nContent-Length: "_s, contentLength, "\r\n\r\n"_s));
+
+                co_await connection.awaitableSend(webkitHTML);
+                co_await connection.awaitableSend(Vector<uint8_t>(1000000, ' '));
```
This is already tested elsewhere and doesn't meaningfully add to the test coverage.  Just send webkitHTML and nothing else, and don't check the commit callbacks.

achristensen07 commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -983,6 +983,60 @@ ExceptionOr<void> LocalDOMWindow::postMessage(JSC::JSGlobalObject& lexicalGlobal
     return { };
 }
 
+ExceptionOr<void> LocalDOMWindow::postMessageFromRemote(JSC::JSGlobalObject& lexicalGlobalObject, LocalDOMWindow& incumbentWindow, std::optional<WebCore::SecurityOriginData> target, const WebCore::MessageWithMessagePorts& message)
```
I don't think there's a reason to pass the LocalDOMWindow& parameter, which is always the same as *this.

achristensen07 commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -983,6 +983,60 @@ ExceptionOr<void> LocalDOMWindow::postMessage(JSC::JSGlobalObject& lexicalGlobal
+    RefPtr<ScriptCallStack> stackTrace;
+    if (InspectorInstrumentation::consoleAgentEnabled(sourceDocument.get()))
+        stackTrace = createScriptCallStack(JSExecState::currentState());
+    auto postMessageIdentifier = InspectorInstrumentation::willPostMessage(*frame());
```
null check frame.

achristensen07 commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -983,6 +983,60 @@ ExceptionOr<void> LocalDOMWindow::postMessage(JSC::JSGlobalObject& lexicalGlobal
+        if (!isCurrentlyDisplayedInFrame())
+            return;
+
+        Ref frame = *this->frame();
```
null check frame

achristensen07 commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -983,6 +983,60 @@ ExceptionOr<void> LocalDOMWindow::postMessage(JSC::JSGlobalObject& lexicalGlobal
+        if (target) {
+            auto targetOrigin = target->securityOrigin();
+            // Check target origin now since the target document may have changed since the timer was scheduled.
+            if (!targetOrigin->isSameSchemeHostPort(document()->securityOrigin())) {
```
This is important to have a test for.  Could you add a test where we try to post a message to an origin that is not the right origin and verify that it is not delivered?

achristensen07 commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -983,6 +983,69 @@ ExceptionOr<void> LocalDOMWindow::postMessage(JSC::JSGlobalObject& lexicalGlobal
+            }
+        }
+
+        if (!document())
```
This should be before we use document() to check the target origin.

achristensen07 commented on Source/WebCore/page/Page.cpp:
```diff
@@ -266,14 +266,14 @@ static constexpr OptionSet<ActivityState::Flag> pageInitialActivityState()
 }
 
-static Ref<Frame> createMainFrame(Page& page, std::variant<UniqueRef<FrameLoaderClient>, UniqueRef<RemoteFrameClient>>&& client, FrameIdentifier identifier)
+static Ref<Frame> createMainFrame(Page& page, std::variant<UniqueRef<FrameLoaderClient>, UniqueRef<RemoteFrameClient>>&& client, FrameIdentifier identifier, std::optional<ProcessIdentifier> remoteProcessIdentifier)
```
This should probably be a variant of UniqueRef<FrameLoaderClient> or a struct RemoteMainFrameCreationParameters that contains a ProcessIdentifier and a UniqueRef<RemoteFrameClient>.  This is because we should never have a RemoteFrameClient but no ProcessIdentifier

achristensen07 commented on Source/WebCore/page/RemoteDOMWindow.cpp:
```diff
@@ -113,13 +114,41 @@ WindowProxy* RemoteDOMWindow::parent() const
+
+    MessageWithMessagePorts messageWithPorts { messageData.releaseReturnValue(), disentangledPorts.releaseReturnValue() };
+    
+    if (frame()) {
```
if (auto* frame = this->frame())

achristensen07 commented on Source/WebCore/page/RemoteDOMWindow.cpp:
```diff
@@ -113,13 +114,41 @@ WindowProxy* RemoteDOMWindow::parent() const
+        if (!sourceDocument)
+            return;
+        target = sourceDocument->securityOrigin().data();
+    } else if (targetOrigin != "*"_s) {
```
I feel like this can share some code with localdomwindow::postMessage by moving some code to DOMWindow.cpp

chirags27 replied:
there seem to be subtle differences (like raising exception in case of LocalDOMWindow::postMessage, which makes me think that trying to share code here will result in unnecessary complication. Let me know if you strongly prefer that.

achristensen07 replied:
This should raise the same exceptions.  Any differences should be related to the remoteness.  In any case, this is a lot of duplicate code that should be refactored in an elegant way to make it so we don't have duplicate code.  We don't want to make a change in one and forget to make the same change in the other.

achristensen07 commented on Source/WebCore/page/RemoteFrame.h:
```diff
@@ -74,6 +77,8 @@ class RemoteFrame final : public Frame {
     RefPtr<RemoteFrameView> m_view;
     UniqueRef<RemoteFrameClient> m_client;
     Markable<LayerHostingContextIdentifier> m_layerHostingContextIdentifier;
+    std::optional<FrameIdentifier> m_remoteFrameID;
```
This should always be the same as this frame's frame identifier, so just remove this.

achristensen07 commented on Source/WebCore/page/RemoteFrame.h:
```diff
@@ -74,6 +77,8 @@ class RemoteFrame final : public Frame {
     UniqueRef<RemoteFrameClient> m_client;
     Markable<LayerHostingContextIdentifier> m_layerHostingContextIdentifier;
+    std::optional<FrameIdentifier> m_remoteFrameID;
+    std::optional<ProcessIdentifier> m_remoteProcessIdentifier;
```
This should not be optional.

achristensen07 commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -983,6 +983,70 @@ ExceptionOr<void> LocalDOMWindow::postMessage(JSC::JSGlobalObject& lexicalGlobal
     return { };
 }
 
+ExceptionOr<void> LocalDOMWindow::postMessageFromRemote(JSC::JSGlobalObject& lexicalGlobalObject, std::optional<WebCore::SecurityOriginData> target, const WebCore::MessageWithMessagePorts& message)
```
This always returns void.  Its return value is never used.  Make it just return void.

achristensen07 commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -983,6 +983,70 @@ ExceptionOr<void> LocalDOMWindow::postMessage(JSC::JSGlobalObject& lexicalGlobal
+    RefPtr<WindowProxy> incumbentWindowProxy = frame() ? &frame()->windowProxy() : nullptr;
+    auto userGestureToForward = UserGestureIndicator::currentUserGesture();
+    RefPtr<ScriptCallStack> stackTrace;
+    if (InspectorInstrumentation::consoleAgentEnabled(sourceDocument.get()))
```
This also has lots of duplicate code that should be refactored and shared.

achristensen07 commented on Source/WebKit/WebProcess/WebProcess.messages.in:
```diff
@@ -213,4 +213,6 @@ messages -> WebProcess LegacyReceiver NotRefCounted {
     OpenDirectoryCacheInvalidated(WebKit::SandboxExtension::Handle handle, WebKit::SandboxExtension::Handle machBootstrapHandle)
 #endif
+
+    RemotePostMessage(struct WebCore::FrameIdentifier windowIdentifier, std::optional<WebCore::SecurityOriginData> target, struct WebCore::MessageWithMessagePorts message)
```
windowIdentifier should probably be renamed to frameIdentifier.

achristensen07 commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1379,6 +1380,25 @@ void WebProcess::setEnhancedAccessibility(bool flag)
+    if (!webFrame)
+        return;
+
+    auto& domWindow = *(webFrame->coreFrame()->window());
```
A few more null checks and early returns.

achristensen07 commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -979,10 +943,63 @@ ExceptionOr<void> LocalDOMWindow::postMessage(JSC::JSGlobalObject& lexicalGlobal
+    if (!frame())
+        return;
+
+    RefPtr<WindowProxy> incumbentWindowProxy = frame() ? &frame()->windowProxy() : nullptr;
```
No need to null check frame() again if we just did.

achristensen07 commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -893,57 +893,21 @@ ExceptionOr<Storage*> LocalDOMWindow::localStorage()
-ExceptionOr<void> LocalDOMWindow::postMessage(JSC::JSGlobalObject& lexicalGlobalObject, LocalDOMWindow& incumbentWindow, JSC::JSValue messageValue, WindowPostMessageOptions&& options)
+void LocalDOMWindow::processPostMessage(JSC::JSGlobalObject& lexicalGlobalObject, RefPtr<Document>&& sourceDocument, const MessageWithMessagePorts& message, RefPtr<WindowProxy>&& incumbentWindowProxy, RefPtr<SecurityOrigin>&& targetOrigin)
 {
-    if (!isCurrentlyDisplayedInFrame())
```
This was dropped.

achristensen07 commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -278,6 +283,7 @@ class LocalDOMWindow final
     {
         return postMessage(globalObject, incumbentWindow, message, WindowPostMessageOptions { WTFMove(targetOrigin), WTFMove(transfer) });
     }
+    WEBCORE_EXPORT void postMessageFromRemote(JSC::JSGlobalObject&, std::optional<WebCore::SecurityOriginData> target, const WebCore::MessageWithMessagePorts&);
```
postMessageFromRemoteFrame?

achristensen07 commented on Source/WebCore/page/RemoteDOMWindow.cpp:
```diff
@@ -113,13 +114,38 @@ WindowProxy* RemoteDOMWindow::parent() const
+    // Compute the target origin. We need to do this synchronously in order
+    // to generate the SyntaxError exception correctly.
+    std::optional<SecurityOriginData> target;
+    if (targetOrigin == "/"_s) {
```
This is still duplicate code.  Refactor this shared code to DOMWindow.cpp.

achristensen07 commented on Source/WebCore/page/RemoteDOMWindow.cpp:
```diff
@@ -113,13 +114,38 @@ WindowProxy* RemoteDOMWindow::parent() const
+        // It doesn't make sense target a postMessage at an opaque origin
+        // because there's no way to represent an opaque origin in a string.
+        if (target->isOpaque())
+            return; // Exception { SyntaxError };
```
This should still return a SyntaxError, and this whole function should return ExceptionOr<void>, and we should have a test that tries to post a message to a RemoteDOMWindow with an explicit opaque origin and verify that it throws the exception in JS with site isolation enabled like it does with site isolation not enabled.

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -495,7 +495,7 @@ static Vector<UserContentURLPattern> parseAndAllowAccessToCORSDisablingPatterns(
 }
 
-static std::variant<UniqueRef<FrameLoaderClient>, UniqueRef<RemoteFrameClient>> clientForMainFrame(Ref<WebFrame>&& mainFrame, bool local)
+static std::variant<UniqueRef<FrameLoaderClient>, PageConfiguration::RemoteMainFrameCreationParameters> clientForMainFrame(Ref<WebFrame>&& mainFrame, bool local, std::optional<WebCore::ProcessIdentifier> remoteProcessIdentifier)
```
We don't need the local boolean any more.  It is redundant info with the std::optional, which contains nullopt for local main frames and a process identifier for remote frames.

chirags27 replied:
good point

achristensen07 commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -46,6 +46,10 @@
 #include <wtf/WeakPtr.h>
 
+namespace Inspector {
+class ScriptCallStack;
```
This looks unused.

pascoej commented on Source/WebKit/WebProcess/WebProcess.h:
```diff
@@ -109,6 +109,8 @@ class SecurityOriginData;
 struct ServiceWorkerContextData;
 #endif
+
+struct GlobalWindowIdentifier;
```
why?

chirags27 committed:
Make postMessage work for site-isolation
https://bugs.webkit.org/show_bug.cgi?id=253588
rdar://106439413

Reviewed by Alex Christensen.

This change makes postMessage ping pong work by creating the Frame with
the correct identifiers and remote processIDs so that message sending
and reception between cross origin iframes can work with postMessage.

* Source/WebCore/page/DOMWindow.cpp:
(WebCore::DOMWindow::createTargetOriginForPostMessage):
* Source/WebCore/page/DOMWindow.h:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::processPostMessage):
(WebCore::LocalDOMWindow::postMessage):
(WebCore::LocalDOMWindow::postMessageFromRemoteFrame):
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/page/Page.cpp:
(WebCore::createMainFrame):
* Source/WebCore/page/PageConfiguration.cpp:
(WebCore::PageConfiguration::PageConfiguration):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebCore/page/RemoteDOMWindow.cpp:
(WebCore::RemoteDOMWindow::postMessage):
* Source/WebCore/page/RemoteDOMWindow.h:
* Source/WebCore/page/RemoteFrame.cpp:
(WebCore::RemoteFrame::createMainFrame):
(WebCore::RemoteFrame::createSubframe):
(WebCore::RemoteFrame::createSubframeWithContentsInAnotherProcess):
(WebCore::RemoteFrame::RemoteFrame):
(WebCore::m_remoteProcessIdentifier):
(WebCore::m_layerHostingContextIdentifier): Deleted.
* Source/WebCore/page/RemoteFrame.h:
* Source/WebCore/page/RemoteFrameClient.h:
* Source/WebKit/Shared/FrameTreeCreationParameters.h:
* Source/WebKit/Shared/FrameTreeCreationParameters.serialization.in:
* Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
(WebKit::ProvisionalFrameProxy::ProvisionalFrameProxy):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::commitProvisionalFrame):
(WebKit::WebFrameProxy::frameTreeCreationParameters const):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::postMessageToRemote):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::postMessageToRemote):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::createRemoteSubframe):
(WebKit::WebFrame::didCommitLoadInAnotherProcess):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::clientForMainFrame):
(WebKit::m_appHighlightsVisible):
(WebKit::WebPage::constructFrameTree):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::remotePostMessage):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/262334@main
